File: samples/matlab/s007_3d_reconstruction.m
% -----------------------------------------------------------------------
% This file is part of the ASTRA Toolbox
% 
% Copyright: 2010-2016, iMinds-Vision Lab, University of Antwerp
%            2014-2016, CWI, Amsterdam
% License: Open Source under GPLv3
% Contact: astra@uantwerpen.be
% Website: http://www.astra-toolbox.com/
% -----------------------------------------------------------------------
vol_geom = astra_create_vol_geom(128, 128, 128);

angles = linspace2(0, 2*pi, 180);
proj_geom = astra_create_proj_geom('cone', 1.0, 1.0, 128, 192, angles, 512, 0);

% Create a simple hollow cube phantom
cube = zeros(128,128,128);
cube(17:112,17:112,17:112) = 1;
cube(33:96,33:96,33:96) = 0;

cfg = astra_struct('cuda3d');
cfg.ProjectionKernel = 'line';
cfg.ProjectionGeometry = proj_geom;
cfg.VolumeGeometry = vol_geom;
projector_id = astra_mex_projector3d('create', cfg);

proj_id = astra_mex_data3d('create', '-proj3d', proj_geom, 0);
vol_id = astra_mex_data3d('create', '-vol', vol_geom, cube);

cfg = astra_struct('FP3D_CUDA');
cfg.ProjectorId = projector_id;
cfg.ProjectionDataId = proj_id;
cfg.VolumeDataId = vol_id;

alg_id = astra_mex_algorithm('create', cfg);
astra_mex_algorithm('run', alg_id);
astra_mex_algorithm('delete', alg_id);

proj_data = astra_mex_data3d('get_single', proj_id);

[proj_id2, proj_data2] = astra_create_sino3d_cuda(cube, proj_geom, vol_geom);

i = 22;

% Display a single projection image
figure, imshow(squeeze(proj_data(:,i,:))',[])
figure, imshow(squeeze(proj_data2(:,i,:))',[])
figure, imshow(squeeze(proj_data(:,i,:) - proj_data2(:,i,:))',[])

%{
% Create a data object for the reconstruction
rec_id = astra_mex_data3d('create', '-vol', vol_geom);

% Set up the parameters for a reconstruction algorithm using the GPU
cfg = astra_struct('SIRT3D_CUDA');
cfg.ReconstructionDataId = rec_id;
cfg.ProjectionDataId = proj_id;
cfg.ProjectorId = projector_id;


% Create the algorithm object from the configuration structure
alg_id = astra_mex_algorithm('create', cfg);

% Run 150 iterations of the algorithm
% Note that this requires about 750MB of GPU memory, and has a runtime
% in the order of 10 seconds.
tic
astra_mex_algorithm('iterate', alg_id, 150);
toc
% Get the result
rec = astra_mex_data3d('get', rec_id);
%}

W = opTomo('cuda', proj_geom, vol_geom);

tic
rec = lsqr(W, proj_data(:), 1e-4, 100);
toc
rec = reshape(rec, astra_geom_size(vol_geom));

figure, imshow(squeeze(rec(:,:,65)),[]);


% Clean up. Note that GPU memory is tied up in the algorithm object,
% and main RAM in the data objects.
%astra_mex_algorithm('delete', alg_id);
astra_mex_projector3d('delete', projector_id);
%astra_mex_data3d('delete', rec_id);
astra_mex_data3d('delete', proj_id);
astra_mex_data3d('delete', vol_id);

File: samples/python/s007_3d_reconstruction.py
# -----------------------------------------------------------------------
# Copyright: 2010-2018, imec Vision Lab, University of Antwerp
#            2013-2018, CWI, Amsterdam
#
# Contact: astra@astra-toolbox.com
# Website: http://www.astra-toolbox.com/
#
# This file is part of the ASTRA Toolbox.
#
#
# The ASTRA Toolbox is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# The ASTRA Toolbox is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with the ASTRA Toolbox. If not, see <http://www.gnu.org/licenses/>.
#
# -----------------------------------------------------------------------

import astra
import numpy as np

vol_geom = astra.create_vol_geom(128, 128, 128)

angles = np.linspace(0, 2*np.pi, 180,False)
proj_geom = astra.create_proj_geom('cone', 1.0, 1.0, 128, 192, angles, 2000, 0)

# Create a simple hollow cube phantom
cube = np.zeros((128,128,128))
cube[17:113,17:113,17:113] = 1
cube[33:97,33:97,33:97] = 0

cfg = astra.astra_dict('cuda3d')
cfg['ProjectionKernel'] = 'line'
cfg['ProjectionGeometry'] = proj_geom
cfg['VolumeGeometry'] = vol_geom
projector_id = astra.projector3d.create(cfg)

# Create projection data from this
proj_id, proj_data = astra.create_sino3d_gpu(cube, proj_geom, vol_geom)

# Display a single projection image
import pylab
pylab.gray()
pylab.figure(1)
pylab.imshow(proj_data[:,20,:])

# Create a data object for the reconstruction
rec_id = astra.data3d.create('-vol', vol_geom)

# Set up the parameters for a reconstruction algorithm using the GPU
astra.plugin.register(astra.plugins.SIRTPlugin)
cfg = astra.astra_dict('SIRT-PLUGIN')
cfg['ReconstructionDataId'] = rec_id
cfg['ProjectionDataId'] = proj_id
cfg['ProjectorId'] = projector_id



# Create the algorithm object from the configuration structure
alg_id = astra.algorithm.create(cfg)

# Run 150 iterations of the algorithm
# Note that this requires about 750MB of GPU memory, and has a runtime
# in the order of 10 seconds.
astra.algorithm.run(alg_id, 150)

# Get the result
rec = astra.data3d.get(rec_id)
pylab.figure(2)
pylab.imshow(rec[:,:,65])
pylab.show()


# Clean up. Note that GPU memory is tied up in the algorithm object,
# and main RAM in the data objects.
astra.algorithm.delete(alg_id)
astra.data3d.delete(rec_id)
astra.data3d.delete(proj_id)
